File: solutions/java/calculator-conundrum/1/src/main/java/CalculatorConundrum.java
import java.util.Objects;

class CalculatorConundrum {
    public String calculate(int operand1, int operand2, String operation) {
        validateOperation(operation);

        return switch (operation) {
            case "+" -> add(operand1, operand2);
            case "*" -> multiply(operand1, operand2);
            case "/" -> divide(operand1, operand2);
            default -> throw new IllegalOperationException(String.format("Operation '%s' does not exist", operation));
        };
    }

    private String add(int a, int b) {
        int result = a + b;

        return buildResult(
                a, b, "+", result);
    }

    private String multiply(int a, int b) {
        int result = a * b;

        return buildResult(
                a, b, "*", result);
    }

    private String divide(int a, int b) {

        if (b == 0) {
            throw new IllegalOperationException("Division by zero is not allowed", new ArithmeticException());
        }

        int result = a / b;

        return buildResult(
                a, b, "/", result);
    }

    private void validateOperation(String operation) {
        if (Objects.isNull(operation)) {
            throw new IllegalArgumentException("Operation cannot be null");
        } else if (operation.isEmpty()) {
            throw new IllegalArgumentException("Operation cannot be empty");
        }
    }

    private String buildResult(int operand1, int operand2, String operation, int result) {
        return String.format(
                "%d %s %d = %d",
                operand1,
                operation,
                operand2,
                result);
    }
}

File: solutions/csharp/calculator-conundrum/1/CalculatorConundrum.cs
enum Operator
{
    Addition,
    Multiplication,
    Division
}

static class OperatorExtensions
{
    public static Operator FromString(string? op)
    {
        return op switch
        {
            "+" => Operator.Addition,
            "*" => Operator.Multiplication,
            "/" => Operator.Division,
            null => throw new ArgumentNullException(nameof(op), "Operation cannot be null"),
            "" => throw new ArgumentException($"Unknown operator {op}", nameof(op)),
            _ => throw new ArgumentOutOfRangeException(nameof(op), op, null)
        };
    }

    public static int Apply(this Operator op, int operand1, int operand2)
    {
        return op switch
        {
            Operator.Addition => operand1 + operand2,
            Operator.Multiplication => operand1 * operand2,
            Operator.Division => operand2 != 0 ? operand1 / operand2 : throw new DivideByZeroException("Division by zero"),
            _ => throw new ArgumentOutOfRangeException(nameof(op), op, null)
        };  
    }
}


public static class SimpleCalculator
{

    private static readonly string[] AllowedOperations = ["+", "*", "/"];
    
    public static string Calculate(int operand1, int operand2, string? operation)
    {
        var op = OperatorExtensions.FromString(operation);
        try
        {
            var value = op.Apply(operand1, operand2);
            return $"{operand1} {operation} {operand2} = {value}";    
        }
        catch (DivideByZeroException)
        {
            return "Division by zero is not allowed.";
        }
    }
}
